File: src/test/java/com/uwetrottmann/trakt5/services/ShowsTest.java
/*
 * Copyright 2024 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.uwetrottmann.trakt5.services;

import com.uwetrottmann.trakt5.BaseTestCase;
import com.uwetrottmann.trakt5.TestData;
import com.uwetrottmann.trakt5.entities.BaseEpisode;
import com.uwetrottmann.trakt5.entities.BaseSeason;
import com.uwetrottmann.trakt5.entities.BaseShow;
import com.uwetrottmann.trakt5.entities.Comment;
import com.uwetrottmann.trakt5.entities.Credits;
import com.uwetrottmann.trakt5.entities.Ratings;
import com.uwetrottmann.trakt5.entities.Show;
import com.uwetrottmann.trakt5.entities.Stats;
import com.uwetrottmann.trakt5.entities.Translation;
import com.uwetrottmann.trakt5.entities.TrendingShow;
import com.uwetrottmann.trakt5.enums.Extended;
import com.uwetrottmann.trakt5.enums.Type;
import org.junit.Test;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class ShowsTest extends BaseTestCase {

    @Test
    public void test_popular() throws IOException {
        List<Show> shows = executeCall(getTrakt().shows().popular(2, null, null));
        assertThat(shows).isNotNull();
        assertThat(shows.size()).isLessThanOrEqualTo(DEFAULT_PAGE_SIZE);
        for (Show show : shows) {
            assertShowNotNull(show);
        }
    }

    @Test
    public void test_trending() throws IOException {
        List<TrendingShow> shows = executeCall(getTrakt().shows().trending(1, null, null));
        assertThat(shows).isNotNull();
        assertThat(shows.size()).isLessThanOrEqualTo(DEFAULT_PAGE_SIZE);
        for (TrendingShow show : shows) {
            assertThat(show.watchers).isNotNull();
            assertShowNotNull(show.show);
        }
    }

    private void assertShowNotNull(Show show) {
        assertThat(show).isNotNull();
        assertThat(show.title).isNotEmpty();
        assertThat(show.ids).isNotNull();
        assertThat(show.ids.trakt).isNotNull();
        assertThat(show.year).isNotNull();
    }

    @Test
    public void test_summary_slug() throws IOException {
        Show show = executeCall(getTrakt().shows().summary(TestData.SHOW_SLUG, Extended.FULL));
        assertTestShow(show);
    }

    @Test
    public void test_summary_trakt_id() throws IOException {
        Show show = executeCall(
                getTrakt().shows().summary(String.valueOf(TestData.SHOW_TRAKT_ID), Extended.FULL));
        assertTestShow(show);
    }

    private void assertTestShow(Show show) {
        assertThat(show).isNotNull();
        assertThat(show.title).isEqualTo(TestData.SHOW_TITLE);
        assertThat(show.year).isEqualTo(TestData.SHOW_YEAR);
        assertThat(show.ids).isNotNull();
        assertThat(show.ids.trakt).isEqualTo(TestData.SHOW_TRAKT_ID);
        assertThat(show.ids.slug).isEqualTo(TestData.SHOW_SLUG);
        assertThat(show.ids.imdb).isEqualTo(TestData.SHOW_IMDB_ID);
        assertThat(show.ids.tmdb).isEqualTo(TestData.SHOW_TMDB_ID);
        assertThat(show.ids.tvdb).isEqualTo(TestData.SHOW_TVDB_ID);
    }

    @Test
    public void test_translations() throws IOException {
        List<Translation> translations = executeCall(getTrakt().shows().translations("breaking-bad"));
        assertThat(translations).isNotNull();
        for (Translation translation : translations) {
            assertThat(translation.language).isNotEmpty();
        }
    }

    @Test
    public void test_translation() throws IOException {
        List<Translation> translations = executeCall(getTrakt().shows().translation("breaking-bad", "de"));
        // we know that Breaking Bad has a German translation, otherwise this test would fail
        assertThat(translations).isNotNull();
        assertThat(translations).hasSize(1);
        assertThat(translations.get(0).language).isEqualTo("de");
    }

    @Test
    public void test_comments() throws IOException {
        List<Comment> comments = executeCall(getTrakt().shows().comments(TestData.SHOW_SLUG, 1, null,
                null));
        assertThat(comments).isNotNull();
        assertThat(comments.size()).isLessThanOrEqualTo(DEFAULT_PAGE_SIZE);
    }

    @Test
    public void test_people() throws IOException {
        Credits credits = executeCall(getTrakt().shows().people(TestData.SHOW_SLUG));
        assertCast(credits, Type.PERSON);
        assertCrew(credits, Type.PERSON);
    }

    @Test
    public void test_ratings() throws IOException {
        Ratings ratings = executeCall(getTrakt().shows().ratings(TestData.SHOW_SLUG));
        assertRatings(ratings);
    }

    @Test
    public void test_stats() throws IOException {
        Stats stats = executeCall(getTrakt().shows().stats(TestData.SHOW_SLUG));
        assertShowStats(stats);
    }

    @Test
    public void test_collected_progress() throws IOException {
        BaseShow show = executeCall(getTrakt()
                .shows()
                .collectedProgress(TestData.SHOW_SLUG, null, null, null, null, null)
        );

        assertThat(show).isNotNull();
        assertThat(show.last_collected_at).isNotNull();
        assertProgress(show);
    }

    @Test
    public void test_watched_progress() throws IOException {
        BaseShow show = executeCall(getTrakt()
                .shows()
                .watchedProgress(TestData.SHOW_SLUG, null, null, null, null, null)
        );

        assertThat(show).isNotNull();
        assertThat(show.last_watched_at).isNotNull();
        assertProgress(show);
    }

    private void assertProgress(BaseShow show) {
        assertThat(show.aired).isGreaterThan(30);
        assertThat(show.completed).isGreaterThanOrEqualTo(1);
        assertThat(show.last_episode).isNotNull();

        // show progress not complete
        assertThat(show.completed).isLessThan(show.aired);
        assertThat(show.next_episode).isNotNull();

        // Killjoys has 5 aired seasons
        assertThat(show.seasons).isNotNull();
        assertThat(show.seasons).hasSize(5);

        BaseSeason season = show.seasons.get(0);
        assertThat(season.number).isEqualTo(1);
        // all aired
        assertThat(season.aired).isEqualTo(10);
        // always at least 1 watched
        assertThat(season.completed).isGreaterThanOrEqualTo(1);

        // episode 1 should always be watched
        assertThat(season.episodes).isNotNull();
        BaseEpisode episode = season.episodes.get(0);
        assertThat(episode.number).isEqualTo(1);
        assertThat(episode.completed).isTrue();
    }

    @Test
    public void test_related() throws IOException {
        List<Show> related = executeCall(getTrakt().shows().related(TestData.SHOW_SLUG, 1, null, null));
        assertThat(related).isNotNull();
        assertThat(related.size()).isLessThanOrEqualTo(DEFAULT_PAGE_SIZE);
        for (Show show : related) {
            assertShowNotNull(show);
        }
    }

}

File: src/test/kotlin/com/redissi/trakt/services/ShowsTest.kt
package com.redissi.trakt.services

import com.redissi.trakt.BaseTestCase
import com.redissi.trakt.TestData
import com.redissi.trakt.entities.BaseShow
import com.redissi.trakt.entities.Show
import com.redissi.trakt.enums.Extended
import com.redissi.trakt.enums.Type
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.*
import org.junit.Test
import java.io.IOException

class ShowsTest : BaseTestCase() {
    @Test
    @Throws(IOException::class)
    fun `get popular shows`() = runBlocking {
        val shows = trakt.shows().popular(2, null, null)
        shows.shouldNotBeNull()
        shows.size.shouldBeLessOrEqualTo(DEFAULT_PAGE_SIZE)
        for (show in shows) {
            assertShowNotNull(show)
        }
    }

    @Test
    @Throws(IOException::class)
    fun `get trending shows`() = runBlocking {
        val trendingShows = trakt.shows().trending(1, null, null)
        trendingShows.shouldNotBeNull()
        trendingShows.size.shouldBeLessOrEqualTo(DEFAULT_PAGE_SIZE)
        for (trendingShow in trendingShows) {
            trendingShow.watchers.shouldNotBeNull()
            trendingShow.show.shouldNotBeNull()
            assertShowNotNull(trendingShow.show!!)
        }
    }

    private fun assertShowNotNull(show: Show) {
        show.shouldNotBeNull()
        show.title.shouldNotBeNullOrEmpty()
        show.ids.shouldNotBeNull()
        show.ids!!.trakt.shouldNotBeNull()
        show.year.shouldNotBeNull()
    }

    @Test
    @Throws(IOException::class)
    fun `get summary for a show from its slug`() = runBlocking {
        val show = trakt.shows().summary(TestData.SHOW_SLUG, Extended.FULL)
        show.shouldNotBeNull()
        assertTestShow(show)
    }

    @Test
    @Throws(IOException::class)
    fun `get summary for a show from its trakt's id`() = runBlocking {
        val show = trakt.shows().summary(TestData.SHOW_TRAKT_ID.toString(), Extended.FULL)
        show.shouldNotBeNull()
        assertTestShow(show)
    }

    private fun assertTestShow(show: Show) {
        show.shouldNotBeNull()
        show.ids.shouldNotBeNull()
        show.title.shouldBeEqualTo(TestData.SHOW_TITLE)
        show.year.shouldBeEqualTo(TestData.SHOW_YEAR)
        show.ids!!.trakt.shouldBeEqualTo(TestData.SHOW_TRAKT_ID)
        show.ids!!.slug.shouldBeEqualTo(TestData.SHOW_SLUG)
        show.ids!!.imdb.shouldBeEqualTo(TestData.SHOW_IMDB_ID)
        show.ids!!.tmdb.shouldBeEqualTo(TestData.SHOW_TMDB_ID)
        show.ids!!.tvdb.shouldBeEqualTo(TestData.SHOW_TVDB_ID)
        show.ids!!.tvrage.shouldBeEqualTo(TestData.SHOW_TVRAGE_ID)
    }

    @Test
    @Throws(IOException::class)
    fun `get translations for a show`() = runBlocking {
        val translations = trakt.shows().translations("breaking-bad")
        translations.shouldNotBeNull()
        for (translation in translations) {
            translation.language.shouldNotBeNullOrEmpty()
        }
    }

    @Test
    @Throws(IOException::class)
    fun `get specific translation for a show`() = runBlocking<Unit> {
        val translations = trakt.shows().translation("breaking-bad", "de")
        // we know that Breaking Bad has a German translation, otherwise this test would fail
        translations.shouldNotBeNull()
        translations.shouldHaveSize(1)
        translations[0].language.shouldBeEqualTo("de")
    }

    @Test
    @Throws(IOException::class)
    fun `get comments for a show`() = runBlocking<Unit> {
        val comments = trakt.shows().comments(TestData.SHOW_SLUG, 1, null,null)
        comments.shouldNotBeNull()
        comments.size.shouldBeLessOrEqualTo(DEFAULT_PAGE_SIZE)
    }

    @Test
    @Throws(IOException::class)
    fun `get credits for a show`() = runBlocking {
        val credits = trakt.shows().people(TestData.SHOW_SLUG)
        credits.shouldNotBeNull()
        assertCast(credits, Type.PERSON)
        assertCrew(credits, Type.PERSON)
    }

    @Test
    @Throws(IOException::class)
    fun `get ratings for a show`() = runBlocking {
        val ratings = trakt.shows().ratings(TestData.SHOW_SLUG)
        ratings.shouldNotBeNull()
        assertRatings(ratings)
    }

    @Test
    @Throws(IOException::class)
    fun `get stats for a show`() = runBlocking {
        val stats = trakt.shows().stats(TestData.SHOW_SLUG)
        stats.shouldNotBeNull()
        assertShowStats(stats)
    }

    @Test
    @Throws(IOException::class)
    fun `get collected progress for a show`() = runBlocking {
        val show = trakt.shows().collectedProgress(TestData.SHOW_SLUG, null, null, null)
        show.shouldNotBeNull()
        show.lastCollectedAt.shouldNotBeNull()
        assertProgress(show)
    }

    @Test
    @Throws(IOException::class)
    fun `get watched progress for a show`() = runBlocking {
        val show = trakt.shows().watchedProgress(TestData.SHOW_SLUG, null, null, null)
        show.shouldNotBeNull()
        show.lastWatchedAt.shouldNotBeNull()
        assertProgress(show)
    }

    private fun assertProgress(show: BaseShow) {
        show.aired.shouldNotBeNull().shouldBeGreaterThan(30)
        show.completed.shouldNotBeNull().shouldBeGreaterOrEqualTo(1)
        // Killjoys has 5 aired seasons
        show.seasons.shouldNotBeNull().shouldHaveSize(5)
        val season = show.seasons!![0]
        season.number.shouldNotBeNull().shouldBe(1)
        // all aired
        season.aired.shouldNotBeNull().shouldBe(10)
        // always at least 1 watched
        season.completed.shouldNotBeNull().shouldBeGreaterOrEqualTo(1)
        // episode 1 should always be watched
        season.episodes.shouldNotBeNull()
        val episode = season.episodes!![0]
        episode.number.shouldNotBeNull().shouldBe(1)
        episode.completed.shouldNotBeNull().shouldBeTrue()
    }

    @Test
    @Throws(IOException::class)
    fun `get related shows`() = runBlocking {
        val related = trakt.shows().related(TestData.SHOW_SLUG, 1, null, null)
        related.shouldNotBeNull()
        related.size.shouldBeLessOrEqualTo(DEFAULT_PAGE_SIZE)
        for (show in related) {
            assertShowNotNull(show)
        }
    }
}